src/io/hdf/hdf_archive.cpp

```diff
@@ -149,6 +150,7 @@ bool hdf_archive::create(const std::filesystem::path& fname, unsigned flags)
 
 bool hdf_archive::open(const std::filesystem::path& fname, unsigned flags)
 {
+  filename_ = fname;
```

Comment:
Add a similar assert.
I also noticed missing wiping out filename_ at `close()`.

Reply:
If there are operations than happen after a file is closed, it would be useful to have the name of the file that was open in the error message.  So clearing the filename_ variable at close doesn't seem helpful.

Reply:
I assumed that there was a 1:1 correspondence between an hdf_archive instance and an HDF file.   But I see from the code that doesn't necessarily have to be the case.  I added a check to hdf_archive, and it turns out there are places in the code where an hfd_archive instance is reused.